### PR TITLE
Extract Aryn token from config for ArynPartitioner

### PR DESCRIPTION
**Pull request by MarkLindblad**
Check `$ARYN_API_KEY` and read `.aryn/config.yaml` in that order to find the Aryn token (Aryn API Key).

### CODE REVIEW

File: lib/sycamore/sycamore/utils/aryn_config.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+_DEFAULT_PATH = os.path.join(pathlib.Path.home(), ".aryn", "config.yaml")
+
+
+class ArynConfig:
+    def __init__(self, **kwargs):
+        self.__dict__.update(kwargs)
+
+
+def get_aryn_api_key(requested_yaml_config_path: str = "") -> str:
```
Comment (alexaryn) [inline lib/sycamore/sycamore/utils/aryn_config.py:14]: `requested_yaml_config_path` is a very long variable name; how about `config_path`?

File: lib/sycamore/sycamore/utils/aryn_config.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    api_key = os.environ.get("ARYN_API_KEY")
+    if api_key:
+        return api_key
+
+    global_aryn_config = get_aryn_config(requested_yaml_config_path)
```
Comment (alexaryn) [inline lib/sycamore/sycamore/utils/aryn_config.py:19]: It looks like we parse the config file every time and throw the result away right after.

File: lib/sycamore/sycamore/utils/aryn_config.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+
+def get_aryn_config(requested_yaml_config_path: str = "") -> ArynConfig:
+    config_path = requested_yaml_config_path or os.environ.get("ARYN_CONFIG") or _DEFAULT_PATH
+    config = ArynConfig()
+    try:
+        with open(config_path, "r") as yaml_file:
+            aryn_env = yaml.safe_load(yaml_file)
+            if aryn_env is None or not isinstance(aryn_env, dict):
+                logging.debug("Aryn YAML config appears to be empty.")
```
Comment (alexaryn) [inline lib/sycamore/sycamore/utils/aryn_config.py:30]: This is probably not "debug" level, but "warning" level.

File: lib/sycamore/sycamore/utils/aryn_config.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
+import os
+import pathlib
+import yaml
+import logging
+
+_DEFAULT_PATH = os.path.join(pathlib.Path.home(), ".aryn", "config.yaml")
+
+
+class ArynConfig:
```
Comment (alexaryn) [inline lib/sycamore/sycamore/utils/aryn_config.py:9]: I think what we really want here is a a singleton class.  It should probably read the config file once and keep the result.  I'd look into using class-variables for this.

File: lib/sycamore/sycamore/utils/aryn_config.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    def __init__(self, **kwargs):
+        self.__dict__.update(kwargs)
```
Comment (alexaryn) [inline lib/sycamore/sycamore/utils/aryn_config.py:11]: Where do we use this cleverness?

I'd worry about the namespace overlap.  What if someone passes a kwarg called 'get_aryn_api_key'?

**Review by alexaryn — COMMENTED** [review]
Looks like a good idea, but I think it could use some refinement.

File: lib/sycamore/sycamore/utils/aryn_config.py (hunk 5 above)
Comment (MarkLindblad) [reply lib/sycamore/sycamore/utils/aryn_config.py]: Removed.

File: lib/sycamore/sycamore/utils/aryn_config.py (hunk 1 above)
Comment (MarkLindblad) [reply lib/sycamore/sycamore/utils/aryn_config.py]: Changed.

File: lib/sycamore/sycamore/utils/aryn_config.py (hunk 4 above)
Comment (MarkLindblad) [reply lib/sycamore/sycamore/utils/aryn_config.py:10]: Changed.

File: lib/sycamore/sycamore/utils/aryn_config.py (hunk 2 above)
Comment (MarkLindblad) [reply lib/sycamore/sycamore/utils/aryn_config.py]: Fixed.

File: lib/sycamore/sycamore/utils/aryn_config.py [hunk 6]
```diff
@@ -2,40 +2,52 @@
...
+        with cls._global_aryn_config_lock:
+            if cls._global_aryn_config:
+                return
+            config_path = config_path or os.environ.get("ARYN_CONFIG") or _DEFAULT_PATH
+
+            try:
+                with open(config_path, "r") as yaml_file:
+                    aryn_config = yaml.safe_load(yaml_file)
+                    cls._global_aryn_config = aryn_config
+                    if aryn_config is None or not isinstance(aryn_config, dict):
```
Comment (alexaryn) [inline lib/sycamore/sycamore/utils/aryn_config.py:43]: None is NoneType so you can just use the isinstance().

File: lib/sycamore/sycamore/utils/aryn_config.py [hunk 7]
```diff
@@ -2,40 +2,52 @@
...
+                        logging.warning("Aryn YAML config appears to be empty.")
+                        return
+                    logging.debug(f"Aryn configuration: {aryn_config}")
+            except FileNotFoundError as err:
+                if config_path == _DEFAULT_PATH:
+                    logging.debug(f"Aryn config YAML not present: {err}")
+                else:
+                    logging.error(f"Unable to load specified aryn config {config_path}: {err}")
```
Comment (alexaryn) [inline lib/sycamore/sycamore/utils/aryn_config.py:51]: I would get rid of the enclosing if-statement and just have this message, but drop the word "specified".

File: lib/sycamore/sycamore/utils/aryn_config.py [hunk 8]
```diff
@@ -2,40 +2,52 @@
...
+            except yaml.scanner.ScannerError as err:
+                logging.error(f"Unable to load {config_path}: {err}. Ignoring the config file.")
```
Comment (alexaryn) [inline lib/sycamore/sycamore/utils/aryn_config.py:53]: "Unable to parse" would be more accurate/useful.  It's debatable that it's valid to continue in this case.

**Review by alexaryn — APPROVED** [review]
Much better.  Just some nits I trust you to address.

File: lib/sycamore/sycamore/utils/aryn_config.py (hunk 6 above)
Comment (MarkLindblad) [reply lib/sycamore/sycamore/utils/aryn_config.py]: Done.

File: lib/sycamore/sycamore/utils/aryn_config.py (hunk 8 above)
Comment (MarkLindblad) [reply lib/sycamore/sycamore/utils/aryn_config.py]: Changed message.

File: lib/sycamore/sycamore/utils/aryn_config.py (hunk 7 above)
Comment (MarkLindblad) [reply lib/sycamore/sycamore/utils/aryn_config.py]: Done

**Pull request merged**